### ISSUE TITLE
Improve error message in range function

### ISSUE THREAD

**Issue by chadagreene** [JuliaLang/julia#45106]
A Type error in the `length` argument of `range` currently produces this error message: 
```julia
julia> range(0,1,length=1e3)
ERROR: TypeError: in keyword argument length, expected Union{Nothing, Integer}, got a value of type Float64
```
The current error message: 
1. fails to mention that the error is related to the `range` function, 
2. is somewhat cryptic about what exactly the source of the problem is,  
3. offers no guidance for what users can do about the problem, and 
4. is a series of sentence fragments that don't actually form a meaningful sentence. 

In my opinion, **all** error messages should _always_ be written in the following order: 
1. Identify the location of the problem. 
2. Describe the technical cause of the problem (in human language). 
3. Conclude with some gentle advice for what can be done about the problem. 

Following the three-step formula above, I propose the following rewrite of the `range` Type error, or something similar to this: 

```julia 
Line 5: ERROR in range(): Keyword argument length was entered as type Float64, but should be of type Integer or Nothing.
```

A discussion of the `range` error message is on the Discourse site [here](https://discourse.julialang.org/t/why-does-scientific-notation-break-the-range-function/80014).

**Comment by stevengj** [comment]
See also [this suggestion](https://discourse.julialang.org/t/why-does-scientific-notation-break-the-range-function/80014/80:
```jl
ERROR: TypeError: in keyword argument length of function range(), expected Union{Nothing, Integer}, got a value of type Float64
```

**Comment by CameronBieganek** [comment]
I think the title of this issue could be made more general. Something like this:

> Improve error messages when a keyword argument type assertion fails.

Here's the generic MWE:

```julia
julia> foo(; x::Union{Nothing, Integer}=nothing) = x
foo (generic function with 1 method)

julia> foo(; x=0.5)
ERROR: TypeError: in keyword argument x, expected Union{Nothing, Integer}, got a value of type Float64
Stacktrace:
 [1] top-level scope
   @ REPL[11]:1
```

It seems to me that there are two main ways to improve this message:

1. Include the function name in the error message. (The function name `foo` does not occur in the stacktrace because the error is technically not happening inside of `foo`.)
2. If the type that is being asserted is not concrete, then change the message to, e.g. `... expected a subtype of Union{Nothing, Integer} ...`.

**Comment by chadagreene** [comment]
@CameronBieganek Is there a way to say `Union{Nothing, Integer}` in human language?  

The function of the error message is to help _people_ understand the problem, whereas `Union{Nothing, Integer}` adds a layer of encryption that will only confuse more and more users, as Julia gets adopted beyond its core group of developers.

There's a real opportunity here to make these error messages more humane.

**Comment by chadagreene** [comment]
I also want to point out the _perspective_ from which the current error message is written: The current message is written from the perspective of the code itself, and it reads as if the code is a narcissist who can only think about its own wants and expectations. The code expected one thing, but got another, boo-hoo. 

A more constructive phrasing would be written for the human reader. With this in mind, please take another look at my suggested wording: It's written for the humans, to help them understand the problem and how to fix it. The difference is subtle, but I think there's a meaningful difference in tone.

**Comment by CameronBieganek** [comment]
It could be done, but it would get a little awkward in the general case. Consider this type:

```julia
Union{Nothing, Integer, String, AbstractVector{<:AbstractFloat}}
```

That would have to be translated to something like this:

```
... expected a value of type Nothing or String, or a value of a type that is a
subtype of Integer or AbstractVector{<:AbstractFloat} ...
```

I guess that's not horrible. :)

"value of a type that is a subtype of" seems a little awkward, but I can't think of a better way to say it that is actually precise.

**Comment by CameronBieganek** [comment]
It seems like Base Julia doesn't generally do this, but I think it would also help with clarity to put backticks around the argument name in the error message, like this:

```
ERROR: TypeError: in keyword argument `length`, expected Union{Nothing, Integer} ...
```

In other words, does the phrase `keyword argument length` mean "the length of a keyword argument", or does it mean "the keyword argument named `length`" ?

**Comment by doriantsolak** [comment]
> Is there a way to say Union{Nothing, Integer} in human language?

I totally get that his causes some confusion if one does not know what a union is. However, I still believe the change should be more minimal. Two points on this:

1. `TypeError` is currently removed in your suggestion and replaced by the function name. However, this is extremely information and I can believe the function name can be printed in reference to the keyword argument. Personally I feel `TypeError` and `MethodError` are some of the most useful and vital information you can receive when Julia throws an error at you.

2. The information that the function expects `Union{Nothing, Integer}` similarly is very vital information and very clear, even to people who have spend just a little time using Julia I would say. Omitting this information and replacing it by very ambiguous non-technical wording like 
> but should be of type Integer or Nothing.
again removes information that is actually presented very clearly and concisely to the user.

Just like in the discourse thread I would again like to propose as a minimal change:

`ERROR: TypeError: in keyword argument length of function range(), expected Union{Nothing, Integer}, got a value of type Float64`

I have spend 10 minutes staring at my screen thinking of something to do with the Union, the best one I would come up with would be:

`ERROR: TypeError: in keyword argument length of function range(), expected value of type Nothing or Integer (Union{Nothing, Integer}), got a value of type Float64`

However, this seems extremely redundant and I personally do not like it. I think some people have said before that Union types are very intuitive and easy to understand. I think that is something I can get behind too, which is why I would not favour this second, longer proposal too much. It is basically an attempt to explain union types in the error message, but then one really has to ask what error messages as its core should do and should not do (so another discussion probably).

**Comment by CameronBieganek** [comment]
The current message is already misleading and unprecise:

```
... expected Union{Nothing, Integer}, got a value of type Float64
```

Note that it is asymmetrical. After the comma it says "value of type Y", but before the comma it just says "X", where X and Y are two different types. It should actually say the following to be precise:

```
expected a value of a type that is a subtype of Union{Nothing, Integer}, got a value of type Float64
```

**Comment by doriantsolak** [comment]
> The current error message implies that you should literally type `foo(; x=Union{Nothing, Integer})`.

Totally agree that `...expected a value of a subtype of Union{Nothing, Integer}...` would be even better. I somehow just wanted to say: `Union{Nothing, Integer}` is useful information, it should be kept.

Being more precise about it is super good 👍

**Comment by StevenWhitaker** [comment]
> ```
> expected a value of a type that is a subtype of Union{Nothing, Integer}
> ```

To me this is starting to feel a bit convoluted (lots of clauses/prepositional phrases). Something like
```
expected `x isa Union{Nothing, Integer}` to hold
```
could be a more succinct, yet still technically correct, alternative.

**Comment by CameronBieganek** [comment]
Yeah, I agree that getting the wording correct and understandable is tricky when talking about values, types, and subtypes...

Equivalently to `x isa Union{Nothing, Integer}`, we could say `typeof(x) <: Union{Nothing, Integer}`, which I think might be easier to understand for beginners.

**Comment by KronosTheLate** [comment]
`typeof(x) <: Union{Nothing, Integer}` is understandable, but not exactly easy to read to not-so-technical users.

**Comment by chadagreene** [comment]
> ...we could say `typeof(x) <: Union{Nothing, Integer}`, which I think might be easier to understand for beginners.

I disagree. A good litmus test for any error message is to read it aloud to a non-expert, and ask them what it means. So try reading this aloud to your housemate or spouse: 

_Can you try to explain the phrase, "Type of less than colon union nothing integer?"_

A much more helpful approach would be to use human language to identify where the problem is, what caused it, and offer guidance on how to fix it. For example, something like this: 

```julia
Line 5: TypeError in range(): Keyword argument length was entered as type Float64, but should be of type Integer or Nothing.
```

**Comment by StevenWhitaker** [comment]
> A good litmus test for any error message is to read it aloud to a non-expert, and ask them what it means.

While I agree that error messages should be understandable by non-experts, I also think that when using a technical tool there needs to be an expectation that users will take time to learn the basics of the tool, at least to some extent. I would consider types to be a basic concept of Julia. Plus, `help?> <:` and `help?> Union` provide decent documentation (though a non-technical user might want to suggest how they can be improved).

Furthermore, there is a tradeoff between human-readability and technical correctness. In this `range` example it might not seem so bad, but it gets more complicated when nested types are involved. For example, if a `Vector{<:Integer}` is required, should the error message just say `input should be of type Vector{Integer}` to avoid `<:` because that can be considered technically complicated? But then the error message would be misleading, because a `Vector{Int}` would work just as well (and would be more performant).

Another benefit of a technically correct error message like `typeof(x) <: Union{Nothing, Integer}` is that it gives users a tool to check for themselves if their input will work, in addition to giving them things to look up so they can better understand the tool they are using.
